Remotion/ObjectBinding/Sample/SessionStateReflectionBusinessObjectStorageProvider.cs

```diff
@@ -0,0 +1,86 @@
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the 
+// GNU Lesser General Public License for more details.
+// 
+// You should have received a copy of the GNU Lesser General Public License
+// along with re-motion; if not, see http://www.gnu.org/licenses.
+// 
+using System;
+using System.Collections.Generic;
+using System.IO;
+using Remotion.Utilities;
+using Remotion.Web.Infrastructure;
+
+namespace Remotion.ObjectBinding.Sample
+{
+  public class SessionStateReflectionBusinessObjectStorageProvider : IReflectionBusinessObjectStorageProvider
+  {
+    private static readonly string s_sessionKeyForStorageProvider = typeof (SessionStateReflectionBusinessObjectStorageProvider).FullName;
+
+    private readonly IHttpContextProvider _httpContextProvider;
+    private readonly IReflectionBusinessObjectStorageProviderFactory _reflectionBusinessObjectStorageProviderFactory;
+
+    public SessionStateReflectionBusinessObjectStorageProvider (
+        IHttpContextProvider httpContextProvider,
+        IReflectionBusinessObjectStorageProviderFactory reflectionBusinessObjectStorageProviderFactory)
+    {
+      ArgumentUtility.CheckNotNull ("httpContextProvider", httpContextProvider);
+      ArgumentUtility.CheckNotNull ("reflectionBusinessObjectStorageProviderFactory", reflectionBusinessObjectStorageProviderFactory);
+
+      _httpContextProvider = httpContextProvider;
+      _reflectionBusinessObjectStorageProviderFactory = reflectionBusinessObjectStorageProviderFactory;
+    }
+
+    /// <inheritdoc />
+    public IReadOnlyCollection<Guid> GetObjectIDsForType (Type type)
+    {
+      ArgumentUtility.CheckNotNull ("type", type);
+
+      var reflectionBusinessObjectStorageProvider = GetReflectionBusinessObjectStorageProviderForCurrentSession();
+      return reflectionBusinessObjectStorageProvider.GetObjectIDsForType (type);
+    }
+
+    /// <inheritdoc />
+    public Stream GetReadObjectStream (Type type, Guid id)
+    {
+      ArgumentUtility.CheckNotNull ("type", type);
+
+      var reflectionBusinessObjectStorageProvider = GetReflectionBusinessObjectStorageProviderForCurrentSession();
+      return reflectionBusinessObjectStorageProvider.GetReadObjectStream (type, id);
+    }
+
+    /// <inheritdoc />
+    public Stream GetWriteObjectStream (Type type, Guid id)
+    {
+      ArgumentUtility.CheckNotNull ("type", type);
+
+      var reflectionBusinessObjectStorageProvider = GetReflectionBusinessObjectStorageProviderForCurrentSession();
+      return reflectionBusinessObjectStorageProvider.GetWriteObjectStream (type, id);
+    }
+
+    private IReflectionBusinessObjectStorageProvider GetReflectionBusinessObjectStorageProviderForCurrentSession ()
+    {
+      var httpContext = _httpContextProvider.GetCurrentHttpContext();
+      var httpSession = httpContext.Session;
+
+      var reflectionBusinessObjectStorageProvider = (IReflectionBusinessObjectStorageProvider) httpSession?[s_sessionKeyForStorageProvider];
```

Comment:
since the null-coalese doesn't work, would it be better to do a session-null-check up front and return the `_reflectionBusinessObjectStorageProviderFactory.CreateBusinessObjectStorageProvider();` before continuing with the "main" code?

Yes, we would do the Create-call twice. On the other hand, we would make it clear that the first call is a special case instead of just a rather obscure guard (hey I'm not a bug, please believe me!) in the main code?

Thoughts?

Reply:
It is probably a bit cleaner if we check for null beforehand. My biggest issue with this code parts is that it is marked as "heuristically unreachable" because ReSharper/VisualStudio thinks that Session cannot be null. But sadly, we cannot get rid of this issue.

Reply:
That should be solved once the NRT annoations are here.